### PR TITLE
Drop route service namespace

### LINKED ISSUE #7204: No validation on create / update that the route's service is within the same namespace
The logic under the covers is making the assumption that the service referenced in the To field is in the namespace, but nothing in the validation actually prevents you from posting an object that points to another namespace.  The service lookup happens against the route's namespace no matter what you specify.

### PR DESCRIPTION
**Pull request by liggitt**
Fixes #7204

### CODE REVIEW

File: pkg/route/registry/route/strategy.go [hunk 1]
```diff
@@ -43,6 +43,8 @@ func (routeStrategy) NamespaceScoped() bool {
 func (s routeStrategy) PrepareForCreate(obj runtime.Object) {
 	route := obj.(*api.Route)
 	route.Status = api.RouteStatus{}
+	// Do not allow cross-namespace routes
+	route.Spec.To.Namespace = ""
```
Comment (deads2k) [inline pkg/route/registry/route/strategy.go]: Doesn't this scream `LocalObjectReference` to you?

Comment (liggitt) [inline pkg/route/registry/route/strategy.go]: LocalObjectReference doesn't have a kind, which I assume we'd want to be able to set eventually... could make it one for now, and expand later?

Comment (deads2k) [inline pkg/route/registry/route/strategy.go]: > LocalObjectReference doesn't have a kind, which I assume we'd want to be able to set eventually... could make it one for now, and expand later?

Let's add a kind.  That ought to be pretty reasonable upstream right?

Comment (liggitt) [inline pkg/route/registry/route/strategy.go]: I'll do that, but that'll be queued behind the codecgen changes I have. I'd rather get this in now and drop it when I convert to a LocalObjectReference

Comment (deads2k) [inline pkg/route/registry/route/strategy.go]: > I'll do that, but that'll be queued behind the codecgen changes I have. I'd rather get this in now and drop it when I convert to a LocalObjectReference

Ok.  Open an issue please.

**Comment by deads2k** [comment]
lgtm pending issue to change the `ObjectReference` to a `LocalObjectReference`

File: pkg/route/api/v1/defaults.go [hunk 2]
```diff
@@ -0,0 +1,29 @@
+package v1
+
+import "k8s.io/kubernetes/pkg/runtime"
+
+func addDefaultingFuncs(scheme *runtime.Scheme) {
+	err := scheme.AddDefaultingFuncs(
+		func(obj *RouteSpec) {
+			if len(obj.To.Kind) == 0 {
```
Comment (liggitt) [inline pkg/route/api/v1/defaults.go]: defaulting is conditional now

File: pkg/route/api/v1beta3/defaults.go [hunk 3]
```diff
@@ -0,0 +1,29 @@
+package v1beta3
+
+import "k8s.io/kubernetes/pkg/runtime"
+
+func addDefaultingFuncs(scheme *runtime.Scheme) {
+	err := scheme.AddDefaultingFuncs(
+		func(obj *RouteSpec) {
+			if len(obj.To.Kind) == 0 {
```
Comment (liggitt) [inline pkg/route/api/v1beta3/defaults.go]: same here

**Comment by liggitt** [comment]
[test]

**Comment by liggitt** [comment]
Opened https://github.com/openshift/origin/issues/7370

Also added validation for kind, and updated defaulting to only set kind to Service if empty, so we fail validation if values other than "Service" are sent to the API, rather than silently overwriting.

**Comment by liggitt** [comment]
@deads2k still look good to you? These bits changed after your review:
https://github.com/openshift/origin/pull/7212/files#diff-1152bfa99c0917bf00854302f786a2ebR43
https://github.com/openshift/origin/pull/7212/files#diff-13cdc93be4a264426a1f7c00583e996dR48

**Comment by deads2k** [comment]
lgtm [merge]

**Pull request merged**